### PR TITLE
fix: BUG: [distance] request confirmation offline doesn't show TBD

### PR DESCRIPTION
**Pull request by paultsimura**
### Details
- Implement map route re-fetch on going back online;
- Show "TBD" on pending fields;

### Fixed Issues

$ https://github.com/Expensify/App/issues/26537
PROPOSAL: https://github.com/Expensify/App/issues/26537#issuecomment-1703516680

### Tests

1. Go offline
2. Click green plus, request money
3. Enter start and finish addresses
4. Click next
5. Click show more to see all fields
- [ ] Verify that fields `amount` and `distance` are filled as "TBD"

6. Go back online
- [ ] Verify that fields `amount` and `distance` are filled with valid values
- [ ] Verify that no errors appear in the JS console

### Offline tests

1. Go offline
2. Click green plus, request money
3. Enter start and finish addresses
4. Click next
5. Click show more to see all fields
- [ ] Verify that fields `amount` and `distance` are filled as "TBD"

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C01GTK53T8Q/p1693944607891839
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/12595293/7ee2352c-6973-4536-8e7a-8e04d66d08db

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/12595293/0db2fbe7-4550-40d5-8210-ab45644fc407

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/12595293/c7dd9b9b-fb9b-4a97-99c4-259938864f63

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/12595293/044af7a8-5c2c-4cd4-ab21-fb64b665e496

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/12595293/6761946a-6920-47fb-952b-fe00495149ab

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/12595293/7517f811-0f6c-4188-97a1-0e71fe7f2e20

</details>

### CODE REVIEW

**Comment by paultsimura** [comment]
@Ollyws they're kind-of related, but not directly. I will report them, tag you in threads, and let you decide what to do with it

**Comment by paultsimura** [comment]
@Ollyws already done, the corresponding Slack thread is referenced in the PR checklist. Waiting for approval...

File: src/components/DistanceRequest.js
Comment (paultsimura) [reply src/components/DistanceRequest.js:95]: Fixed, thanks for pointing this out @Ollyws .
Demonstrating on both pages:

https://github.com/Expensify/App/assets/12595293/3e0fdbc1-4d79-400b-b1cb-cae8b4d98698

https://github.com/Expensify/App/assets/12595293/99c12346-ece0-451f-a54c-aa9fa51678e9

File: src/components/DistanceRequest.js
Comment (paultsimura) [reply src/components/DistanceRequest.js]: There's always room for improvement. Updated, thanks.

File: src/components/DistanceRequest.js
Comment (paultsimura) [reply src/components/DistanceRequest.js]: Of course. Also, since we are talking about polishing up the code, please check out my new commit – I've refactored some extra lines there.

**Comment by hayata-suenaga** [comment]
@paultsimura does this PR also handles refetching of the route when the device comes back online?

**Comment by paultsimura** [comment]
@hayata-suenaga yes, you can see the result in the attached screen recordings - it includes the route re-fetch.

**Comment by paultsimura** [comment]
@hayata-suenaga @Ollyws could you guys please help with getting this tiny translation expedited?

https://expensify.slack.com/archives/C01GTK53T8Q/p1693944607891839

It'd be nice to have enough time for testing this feature for regression before the merge freeze ends.

**Comment by paultsimura** [comment]
@Ollyws the copies have been updated – this PR is ready for the final review

File: src/components/DistanceRequest.js
Comment (paultsimura) [reply src/components/DistanceRequest.js:102]: This specific change is only for cosmetic purposes to reflect my changes in here: https://github.com/Expensify/App/blob/9ce8d13d918cf6a0de89ae673c21dbc435a605b1/src/components/ConfirmedRoute.js#L47

And **that** fix was related not only to non-existing address. Without my change, going back online on the confirmation screen after entering any address points would crash the mobile app.

https://github.com/Expensify/App/assets/12595293/3cc4e401-5b8a-4daa-8976-982c95e5a135

Comment (paultsimura) [reply src/components/DistanceRequest.js:102]: In this particular case we still did return early in the null cases – 4th and 5th checks: `... || lodashIsNull(waypoint.lat) || lodashIsNull(waypoint.lng)`.
However, maybe the fact that I've changed isNull to isNil now handles this case, because it covers `undefined` as well, not only `null`.

**Comment by paultsimura** [comment]
@Ollyws seems like I found the solution, though it was indeed a bit out of scope. Committed.

https://github.com/Expensify/App/assets/12595293/2d605da1-3b4c-4745-b83c-ac62393c7ea9

**Review by Ollyws — APPROVED** [review]
LGTM

File: src/components/ConfirmedRoute.js
Comment (neil-marcellini) [inline src/components/ConfirmedRoute.js:47]: NAB: can this be `if (!waypoint || !waypoint.lng || !waypoint.lat)`?

File: src/components/DistanceRequest.js
Comment (neil-marcellini) [inline src/components/DistanceRequest.js:96]: NAB: I prefer `isRouteAbsentWithoutErrors`

Comment (neil-marcellini) [reply src/components/DistanceRequest.js:96]: However, we also tend to prefer to name variables without negations so maybe you can refactor this a bit.

File: src/components/MoneyRequestConfirmationList.js
Comment (neil-marcellini) [inline src/components/MoneyRequestConfirmationList.js:474]: Instead let's update where we get the distance request merchant to set only the amount to TBD.

https://github.com/Expensify/App/blob/1bdc5acc12f982b4f6e37bfadf9060ac3d2d3b3c/src/components/MoneyRequestConfirmationList.js#L298

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looking pretty good, but I would like to see a few tweaks to put this more in line with what we had designed. I should have included this screenshot for you earlier, sorry. I'm also going to get a designer to review this PR.

Please make these changes:

- Please render the pending map view without any text
- The distance row should show the amount as TBD, but the rate and unit should be visible

**Comment by neil-marcellini** [comment]
Coming from [this comment](https://github.com/Expensify/App/pull/26836#issuecomment-1713493991) let's check if "Solicitar Por determinar" makes sense in spanish. I think we should prevent capitalizing too.

**Comment by paultsimura** [comment]
@neil-marcellini as the map rendering without text was not initially in the scope of the task, and it does require some changes, I would like to discuss it a bit deeper.

1. We use `BlockingView` for the pending map rendering. This component has the `title` prop required. Am I free to change it to optional with `default = ''` ?

https://github.com/Expensify/App/blob/1df618d6e452ce563fcc6ceac37ac034ff8e10a5/src/components/BlockingViews/BlockingView.js#L22

2. What are the precise dimensions of the icon in the new design? And what should I call the new size? As it looks larger than the current max: `iconSizeSuperLarge: 60`

https://github.com/Expensify/App/blob/4984b39545e49b5780602e17e4adc41f8e1678fb/src/styles/variables.ts#L64-L72

3. Should this change take place on both `DistanceRequest` and `Confirmation` pages? Should we remove both subtitles?
    - "The map will be generated when you go back online"
    - "One moment while we set up the map"

**Comment by paultsimura** [comment]
Also @neil-marcellini this PR https://github.com/Expensify/App/pull/26705 has made it impossible to proceed to the next page while being offline (https://github.com/Expensify/App/commit/ace6a400132d776fcac3398ab1e1d279d41e1302). Is it expected?

image

File: src/components/ConfirmedRoute.js
Comment (pradeepmdk) [reply src/components/ConfirmedRoute.js]: when we have lat/lng as 0 value  also ignore here

Comment (paultsimura) [reply src/components/ConfirmedRoute.js]: Right, @neil-marcellini this approach will treat `lat/lng = 0` as falsy and cause the early return.
Even though it's highly unlikely that someone will choose a point with a latitude or longitude equal exactly to 0, the probability is never 0.

How about this?
```js
if (!waypoint || lodashIsNil(waypoint.lng) || lodashIsNil(waypoint.lat))
```

Comment (pradeepmdk) [reply src/components/ConfirmedRoute.js]: incase waypoint.lng is `undefined`  means lodashIsNil returns false only.

Comment (paultsimura) [reply src/components/ConfirmedRoute.js]: No, `lodashIsNil` returns `false` if the value is nullish, meaning `null` or `undefined`.

Comment (pradeepmdk) [reply src/components/ConfirmedRoute.js]: @paultsimura lodashIsNil(null)  =  true , lodashIsNil(undefined)  =  false.  
>lodashIsNil returns false if the value is nullish, 

that is what i am mention in my comment. ` lodashIsNil(waypoint.lng)` if value is `undefined`  means early return will be ignored here.

`lodashIsNil`  will be return true only value have null.

Comment (paultsimura) [reply src/components/ConfirmedRoute.js]: @pradeepmdk please take a look: https://lodash.com/docs/#isNil

Comment (paultsimura) [reply src/components/ConfirmedRoute.js]: @pradeepmdk please take a look: https://lodash.com/docs/#isNil

Comment (neil-marcellini) [reply src/components/ConfirmedRoute.js]: Ah right I keep forgetting they can be 0. Sounds good.

**Comment by neil-marcellini** [comment]
> @neil-marcellini as the map rendering without text was not initially in the scope of the task, and it does require some changes, I would like to discuss it a bit deeper.

Sure, let's discuss.


> 1. We use `BlockingView` for the pending map rendering. This component has the `title` prop required. Am I free to change it to optional with `default = ''` ?

Good question. Maybe we should create a new PendingMapView component with title and subtitle props defaulting to empty. If they are set we could use the BlockingView as we do now and if they are both unset then display the large map icon? I think it makes sense to have a more specialized component.



> 2\. What are the precise dimensions of the icon in the new design? And what should I call the new size? As it looks larger than the current max: `iconSizeSuperLarge: 60`

I'm not sure actually. Please start with what you think looks good and then you can work with the designers to dial in the right size.


> 3\. Should this change take place on both `DistanceRequest` and `Confirmation` pages? If yes – I would need the design for the `DistanceRequest` page with the desired icon size as well;

No thanks, we want to keep the more descriptive text on the distance request page.



> 6\. Should we remove both subtitles for both states?
>     
>     * "The map will be generated when you go back online"
>     * "One moment while we set up the map"

Let's have only the icon on the confirmation page, and keep that descriptive text on the distance request page.

> P.s. As the requirements have expanded, would it be possible to ask for an extended bonus-eligible period on this PR?

Yes I think that sounds reasonable.

**Comment by neil-marcellini** [comment]
> Also @neil-marcellini this PR #26705 has made it impossible to proceed to the next page while being offline ([ace6a40](https://github.com/Expensify/App/commit/ace6a400132d776fcac3398ab1e1d279d41e1302)). Is it expected?

Yes I think that's expected for now. We added that temporarily until the distance request flow is tested and proven to work while offline.

**Comment by paultsimura** [comment]
@neil-marcellini done, ready for your & design review.

image

File: src/libs/DistanceRequestUtils.js
Comment (neil-marcellini) [inline src/libs/DistanceRequestUtils.js:76]: NAB: I don't really like that we're comparing against 0 here, I would rather add a `hasRoute` param and use that. What if we have identical start and stop waypoints, or is that validation handled earlier?

**Review by neil-marcellini — DISMISSED** [review]
Looks good to me pending C+ and design approval.

**Comment by dannymcclain** [comment]
I think this is looking pretty good, but I'm not able to get the map to update even when coming back online (maybe that's expected right now?).

I also wanted to point out that the spacing on this screen is quite a bit more spread out than in the mockup—you can see the difference in these screenshots: 
image
image

That may not be something that we're addressing with _this_ PR though because I'm pretty sure I saw that being worked on somewhere else.

**Comment by paultsimura** [comment]
> I also wanted to point out that the spacing on this screen is quite a bit more spread out than in the mockup

@dannymcclain check this out, I've increased icon dimensions to `120`:

image

**Review by dannymcclain — DISMISSED** [review]
Ok sorry I think I was initially a little unclear about what I was reviewing 🙈. Apologies! This is looking pretty good to me (screenshot attached). 
image

File: src/components/MapView/PendingMapView.tsx
Comment (neil-marcellini) [inline src/components/MapView/PendingMapView.tsx:26]: NAB: Shouldn't this be a style variable?

**Review by neil-marcellini — DISMISSED** [review]
Looking good

File: src/components/MapView/PendingMapView.tsx
Comment (paultsimura) [reply src/components/MapView/PendingMapView.tsx:26]: I think it should, that's why I asked here:

> What are the precise dimensions of the icon in the new design? And what should I call the new size? As it looks larger than the current max: iconSizeSuperLarge: 60

I will name it `iconSizeUltraLarge`, but in general I think this naming needs some standartization.

**Comment by dannymcclain** [comment]
@Ollyws I think it is totally ok in this context for it to just say "Request".
cc @shawnborton because he designed the original mockup.

**Comment by paultsimura** [comment]
Updated with no "TBD" on the button
@dannymcclain @Ollyws @neil-marcellini 

image
image

**Comment by shawnborton** [comment]
Agree, just "Request" works for me!

File: src/components/MoneyRequestConfirmationList.js
Comment (paultsimura) [reply src/components/MoneyRequestConfirmationList.js]: You're right, at the moment the Distance request doesn't allow splitting the bill. Updated, ready for the final review.

**Comment by paultsimura** [comment]
These conflicts appear with the speed of light.
Updated @Ollyws

**Comment by paultsimura** [comment]
@Ollyws thanks man, won't happen again.

**Review by Ollyws — APPROVED** [review]
LGTM

**Review by neil-marcellini — APPROVED** [review]
Great work everyone, thank you.

**Comment by neil-marcellini** [comment]
I'm commenting for the checklist on [this issue](https://github.com/Expensify/App/issues/27895) since we didn't catch something when merging main. Not a big deal.

**Pull request merged**
